src/common/helpers/snap-builds.js

```diff
@@ -25,43 +25,110 @@
 // };
 
 export const BuildStatusColours = {
+  BLUE: 'blue',
   GREEN: 'green',
   YELLOW: 'yellow',
   RED: 'red',
   GREY: 'grey'
 };
 
+const BuildAndPubState = {
+  'NEVER_BUILT':     createState('Never built', BuildStatusColours.GREY, false, 8),
+  'BUILDING_SOON':   createState('Building soon', BuildStatusColours.GREY, 'ellipsis', 7),
+  'IN_PROGRESS':     createState('In progress', BuildStatusColours.BLUE, 'spinner', 2),
+  'FAILED_TO_BUILD': createState('Failed to build', BuildStatusColours.RED, 'cross', 1),
+  'WONT_PUBLISH':    createState('Built, won\'t be published', BuildStatusColours.GREEN, 'tick_outlined', 6),
+  'PUBLISHING_SOON':    createState('Built, publishing soon', BuildStatusColours.GREY, 'tick', 2),
+  'PUBLISHING_NOW':     createState('Built, publishing now', BuildStatusColours.BLUE, 'tick', 3),
+  'PUBLISHING_FAILED':  createState('Built, failed to publish', BuildStatusColours.RED, 'tick', 4),
+  'PUBLISHED':       createState('Built and published', BuildStatusColours.GREEN, 'tick_filled', 5)
+};
+
+function createState(message, colour, icon, priority) {
+  return {
+    message,
+    colour,
+    icon,
+    priority
+  };
+}
+
 // Based on BuildStatusConstants from LP API
 // https://git.launchpad.net/launchpad/tree/lib/lp/buildmaster/enums.py#n22
 //
-// mapping between build status from LP and red, yellow, green, grey status colours
-const BuildStatusMapping = {
-  'Needs building': BuildStatusColours.YELLOW,
-  'Successfully built': BuildStatusColours.GREEN,
-  'Failed to build': BuildStatusColours.RED,
-  'Dependency wait': BuildStatusColours.RED,
-  'Chroot problem': BuildStatusColours.RED,
-  'Build for superseded Source': BuildStatusColours.RED,
-  'Currently building': BuildStatusColours.YELLOW,
-  'Failed to upload': BuildStatusColours.RED,
-  'Uploading build': BuildStatusColours.YELLOW,
-  'Cancelling build': BuildStatusColours.RED,
-  'Cancelled build': BuildStatusColours.RED
+const LaunchpadBuildStates = {
+  'NEEDSBUILT': 'Needs building',
+  'FULLYBUILT': 'Successfully built',
+  'FAILEDTOBUILD': 'Failed to build',
+  'MANUALDEPWAIT': 'Dependency wait',
+  'CHROOTWAIT': 'Chroot problem',
+  'SUPERSEDED': 'Build for superseded Source',
+  'BUILDING': 'Currently building',
+  'FAILEDTOUPLOAD': 'Failed to upload',
+  'UPLOADING': 'Uploading build',
+  'CANCELLING': 'Cancelling build',
+  'CANCELLED': 'Cancelled build'
 };
 
+const LaunchpadStoreUploadStates = {
+  'UNSCHEDULED': 'Unscheduled',
+  'PENDING': 'Pending',
+  'FAILEDTOUPLOAD': 'Failed to upload',
+  'FAILEDTORELEASE': 'Failed to release to channels',
+  'UPLOADED': 'Uploaded'
+};
+
+function mapBuildAndPublishedStates(buildState, publishState) {
+  switch (buildState) {
+    case LaunchpadBuildStates.NEEDSBUILT:
+      return BuildAndPubState.NEVER_BUILT;
+    case LaunchpadBuildStates.FULLYBUILT:
+    case LaunchpadBuildStates.FAILEDTOUPLOAD:
```

Comment:
No.  `FAILEDTOUPLOAD` in `buildState` means that Launchpad failed to fetch the results of a build from the builder, and in that case it will never get as far as attempting to publish it to the store.  You should just put this in the "failed to build" category.